File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l=0, r=0, maxLen=0;
        HashMap<Character,Integer> map = new HashMap<>();
        while(r < s.length()){
            if(map.containsKey(s.charAt(r))){
                if(map.get(s.charAt(r)) >= l){
                    l = map.get(s.charAt(r)) + 1;
                }
            }
            maxLen = Math.max(maxLen,r-l+1);
            map.put(s.charAt(r),r);
            r++;
        }
        return maxLen;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        #We describe a character set 
        char_set = set()
        #We initiate the max len and left pointer from 0
        max_len, l = 0,0
        #iterating through the string using right pointer
        for r in range(len(s)):
            #In case we get element in the char set
            while s[r] in char_set:
                #we will start removing the elements from the other side
                char_set.remove(s[l])
                #and start incrementing the left pointer
                l += 1
            #Once all duplicates are removed, we start adding new elements
            char_set.add(s[r])
            #Once the removing and adding is done we then compare the max length
            max_len = max(max_len,r-l+1)
        #In the end we return the maximum length of the max set
        return max_len